djangocms_versioning_filer/monkeypatch/models.py

```diff
@@ -30,6 +48,31 @@ def is_file_content_valid_for_revert(version, user):
         )
 
 
+def model_save(func):
+    def inner(self, *args, **kwargs):
+        func(self, *args, **kwargs)
+        if self.__class__ == File:
+            # what should we do now?
+            # maybe this has a subclass, but is being saved as a File instance
+            # anyway. do we need to go check all possible subclasses?
+            pass
+        elif issubclass(self.__class__, File):
+            self._file_type_plugin_name = self.__class__.__name__
+        if self._old_is_public != self.is_public and self.pk:
+            self._move_file()
+            self._old_is_public = self.is_public
+        if self.id and self.grouper and not self.grouper.canonical_file_id:
```

Comment:
I'm pretty sure that this should still work in the Admin save_model method because the Versioning query manager override is on the Admin get_queryset method. Why is this logic here?

Reply:
if we are using file_Admin monkey patch in tests when use this 
file_obj = self.create_file_obj(
                original_filename='test-test.doc',
                folder=self.file.folder,
                grouper=grouper,
                publish=False,
            )
file_obj.grouper.canonical_file_id is empty

Reply:
Ok so why is it empty when using the admin monkey patch vs the model save monkey patch?

It is populated in the model save monkey patch because you are running after the save method has been executed because your monkey patch runs the super() before which means an id does exist. This is why one is working and the other doesn't, you're not doing the same thing so it's not the same. You need to execute the save() then try and add the grouper in the admin monkey patch to get the same result!

Reply:
This monkey patch (model.py) executes the code twice! It runs twice because you execute func() before your code which means that all of the code in models.py save() runs first then your code runs. That's what func() is, its the raw code you're monkey patching.

Reply:
I have reverted back the cahnge to fileadmin now. But still when a draft version is published , it should create published version where the grouper canonical_file_id is updated. But even when we can file_obj.save() the monkey patch is not called.